test/studies/ml/lib/Chai.chpl

```diff
@@ -0,0 +1,828 @@
+
+module Chai {
+    
+    import Tensor as tn;
+    // import IO;
+    module IO {
+        public use IO;
+        public use BinaryIO;
+    }
+
+    use Tensor only Tensor;
+    
+
+    record Dense {
+
+        var outputSize: int;
+        var inputSize_ = -1;
+
+        var bias: Tensor(1);
+        var weights: Tensor(2);
+
+        var biasGrad: Tensor(1);
+        var weightsGrad: Tensor(2);
+
+        var uninitialized = true;
+
+        // proc init=(other: Dense) {
+        //     writeln("init was called");
+        //     this.outputSize = other.outputSize;
+        //     this.bias = other.bias;
+        //     this.weights = other.weights;
+        //     this.biasGrad = other.biasGrad;
+        //     this.weightsGrad = other.weightsGrad;
+        //     this.uninitialized = other.uninitialized;
+        // }
+
+
+        proc init(outputSize: int) {
+            this.outputSize = outputSize;
+
+            bias = new Tensor(1,real); // tn.randn(outputSize);
+            weights = new Tensor(2,real); // tn.randn(outputSize, inputSize);
+
+            biasGrad = new Tensor(1,real);
+            weightsGrad = new Tensor(2,real);
+        }
+
+        proc ref initialize(input: Tensor(1)) {
+            if !uninitialized then tn.err("Dense initialize: already initialized");
+            const (inputSize,) = input.shape;
+            inputSize_ = inputSize;
+            const stddevB = sqrt(2.0 / outputSize);
+            const stddevW = sqrt(2.0 / (inputSize + outputSize));
+            bias = tn.zeros(outputSize); // (tn.randn(outputSize) - 0.5) / outputSize; // tn.randn(this.outputSize); // tn.randn(outputSize,mu=0.0,sigma=stddevB);
+            weights = tn.randn(outputSize,inputSize) ;/// (inputSize ** 2.0); // this.weights = tn.randn(this.outputSize, inputSize,mu=0.0,sigma=stddevW);
+            // bias = tn.randn(outputSize,mu=0.0,sigma=stddevB);
+            // weights = tn.randn(outputSize, inputSize,mu=0.0,sigma=stddevW);
+            biasGrad = tn.zeros(outputSize);
+            weightsGrad = tn.zeros(outputSize, inputSize);
+            uninitialized = false;
+        }
+
+        proc ref forwardProp(input: Tensor(1)): Tensor(1) {
+            if uninitialized then initialize(input);
```

Comment:
The dense layer and softmax layer needs to know the shape of the inputs that it will be receiving in order to initialize the weight and bias tensors. IMO this is better than explicitly giving that information when you initialize the layer, since each change to the hyperparameters may effect the shape in the later layers. In this case, the only parameter needed for `Dense` is the number of outputs, since the input size is computed on the first call to the forwardProp method.

Reply:
> since each change to the hyperparameters may effect the shape in the later layers

Okay, I was not thinking about that, makes sense.